File: scripts/fetch-human-reference-smoke-assets.ts
import { createHash } from "node:crypto";
import { existsSync, readFileSync, statSync } from "node:fs";
import { spawnSync } from "node:child_process";
import { ensureDir, fetchText, parseCsv, pathFromRoot, readText, writeCsv, writeJson } from "./lib";

type ReferenceSpec = {
  reference_id: string;
  assembly: "hg38" | "hg19";
  genome_build: "GRCh38" | "GRCh37";
  source_base_url: string;
  chromosomes: string[];
  genes_covered: string[];
};

type SmokeSampleRow = {
  pair_id: string;
  patient: string;
  sample: string;
  role: string;
  status: string;
  assay: string;
  library_strategy: string;
  library_layout: string;
  platform: string;
  model: string;
  run_accession: string;
  fastq_1: string;
  fastq_2: string;
};

const references: ReferenceSpec[] = [
  {
    reference_id: "ucsc_hg38_chr13_chr17",
    assembly: "hg38",
    genome_build: "GRCh38",
    source_base_url: "https://hgdownload.soe.ucsc.edu/goldenPath/hg38/chromosomes",
    chromosomes: ["chr13", "chr17"],
    genes_covered: ["BRCA2", "BRCA1"]
  },
  {
    reference_id: "ucsc_hg19_chr13_chr17",
    assembly: "hg19",
    genome_build: "GRCh37",
    source_base_url: "https://hgdownload.soe.ucsc.edu/goldenPath/hg19/chromosomes",
    chromosomes: ["chr13", "chr17"],
    genes_covered: ["BRCA2", "BRCA1"]
  }
];

const smokePairId = "seqc2_hcc1395_wes_minimal_smoke";
const referenceRoot = "data/raw/reference/human_reference_smoke";
const smokeRoot = "data/raw/smoke/seqc2_hcc1395_human_reference_smoke";
const resultsDir = "results/human_reference_smoke";

function sh(value: string) {
  return `'${value.replaceAll("'", "'\"'\"'")}'`;
}

function run(command: string) {
  const result = spawnSync("bash", ["-lc", command], {
    cwd: pathFromRoot(""),
    encoding: "utf8",
    maxBuffer: 1024 * 1024 * 20
  });
  if (result.status !== 0) {
    throw new Error(`Command failed (${result.status}): ${command}\n${result.stderr}`);
  }
  return result.stdout;
}

function sha256File(relativePath: string) {
  return createHash("sha256").update(readFileSync(pathFromRoot(relativePath))).digest("hex");
}

function md5File(relativePath: string) {
  return createHash("md5").update(readFileSync(pathFromRoot(relativePath))).digest("hex");
}

function parseMd5s(text: string) {
  const md5s = new Map<string, string>();
  for (const line of text.split(/\r?\n/)) {
    const match = line.trim().match(/^([0-9a-fA-F]{32})\s+(\S+)$/);
    if (match) {
      md5s.set(match[2], match[1].toLowerCase());
    }
  }
  return md5s;
}

function sampleName(row: SmokeSampleRow) {
  return row.sample.replace(`_${row.run_accession}`, "");
}

async function main() {
  ensureDir(pathFromRoot("manifests"));
  ensureDir(pathFromRoot(resultsDir));

  const smokeRows = (parseCsv(readText(pathFromRoot("manifests/raw_smoke_samplesheet.csv"))) as SmokeSampleRow[])
    .filter((row) => row.pair_id === smokePairId)
    .sort((a, b) => a.role.localeCompare(b.role));
  if (smokeRows.length !== 2 || !smokeRows.some((row) => row.role === "tumor") || !smokeRows.some((row) => row.role === "normal")) {
    throw new Error(`Expected tumor and normal rows for ${smokePairId}.`);
  }

  const referenceRows: Record<string, unknown>[] = [];
  const samplesheetRows: Record<string, unknown>[] = [];

  for (const reference of references) {
    const referenceDir = `${referenceRoot}/${reference.reference_id}`;
    const fastaPath = `${referenceDir}/${reference.reference_id}.fa`;
    ensureDir(pathFromRoot(referenceDir));
    ensureDir(pathFromRoot(`${smokeRoot}/${reference.reference_id}/bam`));

    const md5s = parseMd5s(await fetchText(`${reference.source_base_url}/md5sum.txt`));
    const sourceUrls: string[] = [];
    const md5Values: string[] = [];
    const localGzPaths: string[] = [];

    for (const chromosome of reference.chromosomes) {
      const fileName = `${chromosome}.fa.gz`;
      const sourceUrl = `${reference.source_base_url}/${fileName}`;
      const gzPath = `${referenceDir}/${fileName}`;
      sourceUrls.push(sourceUrl);
      localGzPaths.push(gzPath);
      const expectedMd5 = md5s.get(fileName) ?? "";
      md5Values.push(expectedMd5);

      if (!existsSync(pathFromRoot(gzPath))) {
        run(`curl -fsSL ${sh(sourceUrl)} -o ${sh(gzPath)}`);
      }

      if (expectedMd5) {
        const observed = md5File(gzPath);
        if (observed !== expectedMd5) {
          throw new Error(`${gzPath} md5 mismatch: expected ${expectedMd5}, observed ${observed}.`);
        }
      }
    }

    if (!existsSync(pathFromRoot(fastaPath))) {
      run(`gzip -cd ${localGzPaths.map(sh).join(" ")} > ${sh(fastaPath)}`);
    }
    if (!existsSync(pathFromRoot(`${fastaPath}.fai`))) {
      run(`samtools faidx ${sh(fastaPath)}`);
    }
    const fastaSha256 = sha256File(fastaPath);

    referenceRows.push({
      reference_id: reference.reference_id,
      assembly: reference.assembly,
      genome_build: reference.genome_build,
      source: "UCSC Genome Browser per-chromosome FASTA",
      source_base_url: reference.source_base_url,
      chromosomes: reference.chromosomes.join(";"),
      genes_covered: reference.genes_covered.join(";"),
      source_urls: sourceUrls.join(";"),
      source_md5s: md5Values.join(";"),
      md5_status: md5Values.every(Boolean) ? "passed" : "not_available",
      fasta_path: fastaPath,
      fasta_fai_path: `${fastaPath}.fai`,
      fasta_sha256: fastaSha256,
      fasta_size_bytes: statSync(pathFromRoot(fastaPath)).size,
      caveat:
        "Partial human-reference smoke containing chr13 and chr17 only. Validates real-reference alignment mechanics, not whole-genome/full-exome performance."
    });

    for (const row of smokeRows) {
      const run = row.run_accession;
      const sample = sampleName(row);
      const bam = `${smokeRoot}/${reference.reference_id}/bam/${run}.coordinate_sorted.bam`;
      samplesheetRows.push({
        pair_id: row.pair_id,
        patient: row.patient,
        sample,
        role: row.role,
        status: row.status,
        assay: row.assay,
        library_strategy: row.library_strategy,
        library_layout: row.library_layout,
        platform: row.platform,
        model: row.model,
        run_accession: run,
        fastq_1: row.fastq_1,
        fastq_2: row.fastq_2,
        reference_id: reference.reference_id,
        assembly: reference.assembly,
        genome_build: reference.genome_build,
        chromosomes: reference.chromosomes.join(";"),
        genes_covered: reference.genes_covered.join(";"),
        reference_path: fastaPath,
        reference_sha256: fastaSha256,
        aligner: "bwa mem",
        aligner_threads: "2",
        read_group_id: `${run}_${reference.assembly}`,
        read_group_sample: sample,
        read_group_library: `${row.assay}_${row.role}_${reference.assembly}`,
        read_group_platform: "ILLUMINA",
        read_group_platform_unit: run,
        output_bam: bam,
        output_bai: `${bam}.bai`,
        source: "Phase 2A local HCC1395 FASTQ subset aligned to partial UCSC human reference",
        caveat:
          "Partial chr13/chr17 human-reference smoke for reference-build and BAM contract validation only; not full-depth WES/WGS, somatic calling, or HRD evidence."
      });
    }
  }

  await writeCsv(pathFromRoot("manifests/human_reference_smoke_references.csv"), referenceRows);
  await writeCsv(pathFromRoot("manifests/human_reference_smoke_samplesheet.csv"), samplesheetRows);
  await writeJson(pathFromRoot(`${resultsDir}/reference_assets_summary.json`), {
    generatedAt: new Date().toISOString(),
    status: "built",
    referenceCount: referenceRows.length,
    sampleRows: samplesheetRows.length,
    references: referenceRows,
    boundary:
      "Phase 2C uses partial UCSC hg38/hg19 chromosome references for local validation. Full-depth Diana or SEQC2 calling still requires full reference bundles, intervals, known-sites resources, and caller configuration."
  });

  console.log(`Built ${referenceRows.length} partial human-reference smoke bundles and ${samplesheetRows.length} sample rows.`);
}

await main();

File: src/diana_omics/commands/fetch_human_reference_smoke_assets.py
from __future__ import annotations

from ..paths import path_from_root
from ..utils import (
    ensure_dir,
    fetch_text,
    iso_now,
    md5_file,
    parse_csv,
    quote_shell_arg,
    read_text,
    run_command,
    sha256_file,
    write_csv,
    write_json,
)

REFERENCES = [
    {
        "reference_id": "ucsc_hg38_chr13_chr17",
        "assembly": "hg38",
        "genome_build": "GRCh38",
        "source_base_url": "https://hgdownload.soe.ucsc.edu/goldenPath/hg38/chromosomes",
        "chromosomes": ["chr13", "chr17"],
        "genes_covered": ["BRCA2", "BRCA1"],
    },
    {
        "reference_id": "ucsc_hg19_chr13_chr17",
        "assembly": "hg19",
        "genome_build": "GRCh37",
        "source_base_url": "https://hgdownload.soe.ucsc.edu/goldenPath/hg19/chromosomes",
        "chromosomes": ["chr13", "chr17"],
        "genes_covered": ["BRCA2", "BRCA1"],
    },
]
SMOKE_PAIR_ID = "seqc2_hcc1395_wes_minimal_smoke"
REFERENCE_ROOT = "data/raw/reference/human_reference_smoke"
SMOKE_ROOT = "data/raw/smoke/seqc2_hcc1395_human_reference_smoke"
RESULTS_DIR = "results/human_reference_smoke"


def parse_md5s(text: str) -> dict[str, str]:
    md5s = {}
    for line in text.splitlines():
        parts = line.strip().split()
        if len(parts) == 2 and len(parts[0]) == 32:
            md5s[parts[1]] = parts[0].lower()
    return md5s


def sample_name(row: dict[str, str]) -> str:
    return row["sample"].replace(f"_{row['run_accession']}", "")


def main() -> None:
    ensure_dir(path_from_root("manifests"))
    ensure_dir(path_from_root(RESULTS_DIR))
    smoke_rows = sorted(
        [row for row in parse_csv(read_text(path_from_root("manifests/raw_smoke_samplesheet.csv"))) if row["pair_id"] == SMOKE_PAIR_ID],
        key=lambda row: row["role"],
    )
    if (
        len(smoke_rows) != 2
        or not any(row["role"] == "tumor" for row in smoke_rows)
        or not any(row["role"] == "normal" for row in smoke_rows)
    ):
        raise RuntimeError(f"Expected tumor and normal rows for {SMOKE_PAIR_ID}.")

    reference_rows = []
    samplesheet_rows = []
    for reference in REFERENCES:
        reference_dir = f"{REFERENCE_ROOT}/{reference['reference_id']}"
        fasta_path = f"{reference_dir}/{reference['reference_id']}.fa"
        ensure_dir(path_from_root(reference_dir))
        ensure_dir(path_from_root(f"{SMOKE_ROOT}/{reference['reference_id']}/bam"))
        md5s = parse_md5s(fetch_text(f"{reference['source_base_url']}/md5sum.txt"))
        source_urls = []
        md5_values = []
        local_gz_paths = []
        for chromosome in reference["chromosomes"]:
            file_name = f"{chromosome}.fa.gz"
            source_url = f"{reference['source_base_url']}/{file_name}"
            gz_path = f"{reference_dir}/{file_name}"
            source_urls.append(source_url)
            local_gz_paths.append(gz_path)
            expected_md5 = md5s.get(file_name, "")
            md5_values.append(expected_md5)
            if not path_from_root(gz_path).exists():
                run_command(f"curl -fsSL {quote_shell_arg(source_url)} -o {quote_shell_arg(gz_path)}")
            if expected_md5 and md5_file(gz_path) != expected_md5:
                raise RuntimeError(f"{gz_path} md5 mismatch: expected {expected_md5}, observed {md5_file(gz_path)}.")
        if not path_from_root(fasta_path).exists():
            run_command(f"gzip -cd {' '.join(quote_shell_arg(path) for path in local_gz_paths)} > {quote_shell_arg(fasta_path)}")
        if not path_from_root(f"{fasta_path}.fai").exists():
            run_command(f"samtools faidx {quote_shell_arg(fasta_path)}")
        fasta_sha256 = sha256_file(fasta_path)
        reference_rows.append(
            {
                "reference_id": reference["reference_id"],
                "assembly": reference["assembly"],
                "genome_build": reference["genome_build"],
                "source": "UCSC Genome Browser per-chromosome FASTA",
                "source_base_url": reference["source_base_url"],
                "chromosomes": ";".join(reference["chromosomes"]),
                "genes_covered": ";".join(reference["genes_covered"]),
                "source_urls": ";".join(source_urls),
                "source_md5s": ";".join(md5_values),
                "md5_status": "passed" if all(md5_values) else "not_available",
                "fasta_path": fasta_path,
                "fasta_fai_path": f"{fasta_path}.fai",
                "fasta_sha256": fasta_sha256,
                "fasta_size_bytes": path_from_root(fasta_path).stat().st_size,
                "caveat": "Partial human-reference smoke containing chr13 and chr17 only. Validates real-reference alignment mechanics, not whole-genome/full-exome performance.",
            }
        )
        for row in smoke_rows:
            run = row["run_accession"]
            sample = sample_name(row)
            bam = f"{SMOKE_ROOT}/{reference['reference_id']}/bam/{run}.coordinate_sorted.bam"
            samplesheet_rows.append(
                {
                    **{
                        key: row[key]
                        for key in [
                            "pair_id",
                            "patient",
                            "role",
                            "status",
                            "assay",
                            "library_strategy",
                            "library_layout",
                            "platform",
                            "model",
                            "run_accession",
                            "fastq_1",
                            "fastq_2",
                        ]
                    },
                    "sample": sample,
                    "reference_id": reference["reference_id"],
                    "assembly": reference["assembly"],
                    "genome_build": reference["genome_build"],
                    "chromosomes": ";".join(reference["chromosomes"]),
                    "genes_covered": ";".join(reference["genes_covered"]),
                    "reference_path": fasta_path,
                    "reference_sha256": fasta_sha256,
                    "aligner": "bwa mem",
                    "aligner_threads": "2",
                    "read_group_id": f"{run}_{reference['assembly']}",
                    "read_group_sample": sample,
                    "read_group_library": f"{row['assay']}_{row['role']}_{reference['assembly']}",
                    "read_group_platform": "ILLUMINA",
                    "read_group_platform_unit": run,
                    "output_bam": bam,
                    "output_bai": f"{bam}.bai",
                    "source": "Phase 2A local HCC1395 FASTQ subset aligned to partial UCSC human reference",
                    "caveat": "Partial chr13/chr17 human-reference smoke for reference-build and BAM contract validation only; not full-depth WES/WGS, somatic calling, or HRD evidence.",
                }
            )
    write_csv(path_from_root("manifests/human_reference_smoke_references.csv"), reference_rows)
    write_csv(path_from_root("manifests/human_reference_smoke_samplesheet.csv"), samplesheet_rows)
    write_json(
        path_from_root(f"{RESULTS_DIR}/reference_assets_summary.json"),
        {
            "generatedAt": iso_now(),
            "status": "built",
            "referenceCount": len(reference_rows),
            "sampleRows": len(samplesheet_rows),
            "references": reference_rows,
            "boundary": "Phase 2C uses partial UCSC hg38/hg19 chromosome references for local validation. Full-depth Diana or SEQC2 calling still requires full reference bundles, intervals, known-sites resources, and caller configuration.",
        },
    )
    print(f"Built {len(reference_rows)} partial human-reference smoke bundles and {len(samplesheet_rows)} sample rows.")


if __name__ == "__main__":
    main()
